### PR TITLE
Truncate abstract Slate object

### PR DESCRIPTION
**Pull request by oliverroick**
Change #359 introduced a regression because we changed the `abstract` field from plain text to rich text. The underlying data structure changed from `string` to `object` so truncating the abstract with the existing method won't work any longer. 

I first fixed this the simple way. Use Slate's `Node.string` to get the text across all children and truncate that.

Works fine until the first 240 characters include an equation. In that case we would show the latex equivalent of the equation, something like `Lorem ipsum e=mc^2 blah blah`. 

So I went and implemented a way to truncate slate node objects. The code is in this PR. This basically shows the rich text-equivalent of the abstract, including all equations. 

The result may look like this:

Screenshot 2022-05-26 at 12 38 41

This may stretch the grid vertically leaving some ugly white space in neighbouring cells. What are you thoughts on this?

We've got only two options, I think:

- We use the plain text equivalent and don't deal with equations
- We use the current implementation and treat equations in abstracts as edge cases, that don't look nice but also don't break the application.

### CODE REVIEW

File: app/assets/scripts/components/slate/plugins/common/text-utils.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      children: truncatedChildren
+    };
+  }
+
+  if (keys.includes('text')) {
+    const length = value.text.trim().length;
+    if (length + sumLength > maxChars) {
+      const truncateAt = maxChars - sumLength;
+      const slicedText = value.text.slice(0, truncateAt);
+      const truncatedText = slicedText.slice(0, slicedText.lastIndexOf(' '));
```
Comment (danielfdsilva) [inline app/assets/scripts/components/slate/plugins/common/text-utils.js:53]: clever!

**Review by danielfdsilva — CHANGES_REQUESTED** [review]
It seems that the sum length is resetting from between different children.

| source | render |
|---|---|
| image | image |

I think that having tables and equations on the summary will definitely be an edge-case and perhaps we should leave it to the editors to decide.  
However we could also include restrictions on the editor. For this specific field no block level elements would be allowed for example - not sure it is worth the effort though.

File: app/assets/scripts/components/slate/plugins/common/text-utils.js (hunk 1 above)
Comment (oliverroick) [reply app/assets/scripts/components/slate/plugins/common/text-utils.js:65]: Clever indeed. But I stole this from you: https://github.com/NASA-IMPACT/nasa-apt-frontend/blame/develop/app/assets/scripts/components/documents/hub/document-hub-entry.js#L50-L53

**Comment by oliverroick** [comment]
@danielfdsilva Thanks for catching this. I fixed it now.

File: app/assets/scripts/components/slate/plugins/common/text-utils.js (hunk 1 above)
Comment (danielfdsilva) [reply app/assets/scripts/components/slate/plugins/common/text-utils.js:65]: ah. I did not remember that. 🤦

**Review by danielfdsilva — CHANGES_REQUESTED** [review]
Truncate works better now, but I found a problem with the tables.

| source | result |
| -- | -- |
| image | image |

There's a whole cell being filtered out. It seems that the difficulty of truncating such object is not so much a technical one but more conceptual. What would be the best behavior for truncating an equation? How do you properly handle tables?

Since these elements are not really supposed to be used here, I would propose that we strip them out and replace with `[...]`, indicating omission of content. This would be done for equations, tables, images, and headings. Each would be in a paragraph (to simplify implementation) and have some sort of length penalty (like counting as 15 characters).

image

What do you think of this?

**Comment by oliverroick** [comment]
@deborahUAH Because we changed the abstracts to a rich-text field, we also need to adapt how the truncate abstracts in the documents overview. Since abstracts can now contain tables and images, truncating abstracts has become more complex and can lead to strange side effects (see [Daniels comment](https://github.com/NASA-IMPACT/nasa-apt-frontend/pull/364#pullrequestreview-989107650)).

We have two options to deal with these situations.

1. As Daniel outlined, we can replace images and tables with a placeholder, which could affect the readability of truncated abstracts because crucial information might be missing.
2. We restrict the types of content that authors can add to an abstract. Specifically we could not allow tables and images, and in turn we don't have to deal with those when truncating abstracts.  

Option 2, in my opinion, is the more elegant solution. Abstracts in scientific publications usually do not contain images or tables. Authors will still be able to use text formatting (bold, italic, ...) and add equations. 

Do you agree with this approach? Is there anything we haven't considered?

**Comment by deborahUAH** [comment]
Use option 2, I can not think of the need for images or tables in the abstract.  So restrict the types of content that authors can add to an abstract.  Equations are needed however and this where the in-line equations are needed more so than separate line equations.  Symbols are often needed as well.

**Comment by oliverroick** [comment]
@danielfdsilva I've made changes to restrict the abstract field, so authors won't be able to add images, tables, and headers anymore. Can we get this approved?

**Review by danielfdsilva — APPROVED** [review]

**Pull request merged**